lxd/main_forksyscall.go

```diff
@@ -567,6 +567,61 @@ static void mount_emulate(void)
 	}
 }
 
+static bool lxd_cap_is_set(cap_t caps, cap_value_t cap, cap_flag_t flag)
+{
+	int ret;
+	cap_flag_value_t flagval;
+
+	ret = cap_get_flag(caps, cap, flag, &flagval);
+	if (ret < 0)
+		return false;
+
+	return flagval == CAP_SET;
+}
+
+static void sched_setscheduler_emulate(void)
+{
+	__do_close int pidfd = -EBADF, ns_fd = -EBADF;
+	pid_t pid_caller = -ESRCH, pid_target = -ESRCH;
+	int policy = -1, sched_priority = -1;
+	struct sched_param param = {};
+	cap_t caps;
+
+	pid_caller = atoi(advance_arg(true));
+
+	pidfd = atoi(advance_arg(true));
+	ns_fd = pidfd_nsfd(pidfd, pid_caller);
+	if (ns_fd < 0)
+		_exit(EXIT_FAILURE);
+
+	pid_target = atoi(advance_arg(true));
+	if (pid_target < 0)
+		_exit(EXIT_FAILURE);
+
+	policy = atoi(advance_arg(true));
+	if (policy < 0)
+		_exit(EXIT_FAILURE);
+
+	sched_priority = atoi(advance_arg(true));
+	if (sched_priority < 0)
+		_exit(EXIT_FAILURE);
+
+	param.sched_priority = sched_priority;
+
+	caps = cap_get_pid(pid_caller);
+	if (!caps)
+		_exit(EXIT_FAILURE);
+
+	if (!lxd_cap_is_set(caps, CAP_SYS_NICE, CAP_EFFECTIVE))
+		_exit(EXIT_FAILURE);
+
+	if (!change_namespaces(pidfd, ns_fd, CLONE_NEWPID))
+		_exit(EXIT_FAILURE);
+
+	if (sched_setscheduler(pid_target, policy, &param))
```

Comment:
This is another detail I am not sure is correct, or at least I don't see how it can be. `pid_target` as being passed to `lxd forksyscall` is still the pid from outside of the namespace (as being retrieved from `siov.req.pid`). As we have now switched to the target namespace the `pid_target` is no longer valid as it hasn't been translated to the targets PID namespace, right?

Reply:
cc @brauner @stgraber

Reply:
That's only true for the `0` case, i.e. when we hit:
```go
	// If the caller passed zero they want to change their own attributes.
	if args.pidTarget == 0 {
		args.pidTarget = args.pidCaller
	}
```

in the other case we take `pidTarget` from `siov.req.data.args[0]` which is untranslated. This is all racy btw.

Reply:
That's right but things are going to be racy already by the point the Android side selected the pid to target.

Reply:
Yeah, but we're making the race window quite a bit wider with the forksyscall. :)

Reply:
Yes. Is there a reason we have to use `lxd forksyscall` in the `pidTarget == 0` case? As that is the case in the majority of the calls we have from Android it could be a simple win by not forking and do the `sched_setscheduler` call right away.